### PR TITLE
[SPARK-9360][SQL] Support BinaryType in PrefixComparators for UnsafeExternalSort

### PR DESCRIPTION
**Pull request by maropu**
The current implementation of UnsafeExternalSort uses NoOpPrefixComparator for binary-typed data.
So, we need to add BinaryPrefixComparator in PrefixComparators.

### CODE REVIEW

**Comment by sarutak** [comment]
/CC @rxin 
@maropu Could you update this to merge cleanly?

**Comment by maropu** [comment]
thanks, I'll fix it.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java [hunk 1]
```diff
@@ -52,6 +59,38 @@ public int compare(long bPrefix, long aPrefix) {
...
+        return 0L;
+      } else {
+        /**
+         * TODO: If a wrapper for BinaryType is created (SPARK-8786),
+         * these codes below will be in the wrapper class.
+         */
+        final int minLen = Math.min(bytes.length, 8);
+        long p = 0;
+        for (int i = 1; i <= minLen; ++i) {
+          p |= (128L + PlatformDependent.UNSAFE.getByte(bytes, BYTE_ARRAY_OFFSET + i - 1))
```
Comment (rxin) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: can we get a word directly, similar to how we do it in utf8string?

Comment (maropu) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: I think it is hard to do that.
This byte access is needed to map signed bytes to unsigned ones, and 
this mapping can make `UnsignedLongs#compare` compare them in a order-preserving way according to `TypeUtils#compareBinary`.

If we have direct word access here, `BinaryPrefixComparator#compare` needs to inefficiently compare them by using `TypeUtils#compareBinary`.

**Comment by maropu** [comment]
@rxin The tests failed though,  ISTM the ``hive-thriftserver`` tests are not related to this pr.
This failure is correct, or not?

**Comment by rxin** [comment]
cc @davies for review

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java (hunk 1 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: Why use 1 as the initial value? I think use 0 could be better.
```
for (int i=0; i<minLen; i++) {
 p <<= 8;
 p |= 128L + UNSAFE.getByte(bytes, BYTE_ARRAY_OFFSET + i)
}
p <<= (8 - minLen) * 8;
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala [hunk 2]
```diff
@@ -76,6 +78,7 @@ case class SortPrefix(child: SortOrder) extends UnaryExpression {
         (DoublePrefixComparator.computePrefix(Double.NegativeInfinity),
           s"$DoublePrefixCmp.computePrefix((double)$input)")
       case StringType => (0L, s"$input.getPrefix()")
+      case BinaryType => (0L, s"$BinaryPrefixCmp.computePrefix((byte[])$input)")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: I think we don't need the cast here.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala]: fixed

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java (hunk 1 above)
Comment (maropu) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: fixed

**Comment by maropu** [comment]
@davies @rxin ok, all the comments applied.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java [hunk 3]
```diff
@@ -20,15 +20,22 @@
 import com.google.common.primitives.UnsignedLongs;
 
 import org.apache.spark.annotation.Private;
+import org.apache.spark.unsafe.PlatformDependent;
 import org.apache.spark.unsafe.types.UTF8String;
 import org.apache.spark.util.Utils;
 
+import java.nio.ByteOrder;
+
+import static org.apache.spark.unsafe.PlatformDependent.*;
```
Comment (davies) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: These are not used here.

**Comment by davies** [comment]
LGTM, merging this into master and 1.5.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java (hunk 3 above)
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: Actually, they were: they're needed for `BYTE_ARRAY_OFFSET`.

Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: This broke the build, so I'm hotfixing now.

Comment (davies) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: Sorry, just missed that, thanks!

**Pull request closed without merging**